catalog/spatial/registry/registry-api-impl/src/main/java/org/codice/ddf/registry/api/impl/RegistryStoreCleanupHandler.java

```diff
@@ -0,0 +1,157 @@
+ * is removed.
+ */
+public class RegistryStoreCleanupHandler implements EventHandler {
+    private static final Logger LOGGER = LoggerFactory.getLogger(RegistryStoreCleanupHandler.class);
+
+    private FederationAdminService federationAdminService;
+
+    private ExecutorService executor;
+
+    private static final int SHUTDOWN_TIMEOUT_SECONDS = 60;
+
+    private boolean cleanupRelatedMetacards = true;
+
+    private Map<Object, RegistryStore> registryStorePidToServiceMap = new ConcurrentHashMap<>();
+
+    public void destroy() {
+        executor.shutdown();
+        try {
+            if (!executor.awaitTermination(SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
+                executor.shutdownNow();
+                if (!executor.awaitTermination(SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
+                    LOGGER.debug("Thread pool failed to terminate");
+                }
+            }
+        } catch (InterruptedException e) {
+            executor.shutdownNow();
+        }
+    }
+
+    public void bindRegistryStore(ServiceReference serviceReference) {
+        BundleContext bundleContext = getBundleContext();
+
+        if (serviceReference != null && bundleContext != null) {
+            RegistryStore registryStore =
+                    (RegistryStore) bundleContext.getService(serviceReference);
+
+            registryStorePidToServiceMap.put(serviceReference.getProperty(Constants.SERVICE_PID),
+                    registryStore);
+        }
+    }
+
+    public void unbindRegistryStore(ServiceReference serviceReference) {
+        if (serviceReference != null) {
+            registryStorePidToServiceMap.remove(serviceReference.getProperty(Constants.SERVICE_PID));
+        }
+    }
+
+    @Override
+    public void handleEvent(Event event) {
+        if (!cleanupRelatedMetacards) {
+            return;
+        }
+
+        if (((ServiceEvent) event.getProperty(EventConstants.EVENT)).getType()
+                != ServiceEvent.UNREGISTERING) {
+            return;
+        }
+        Object servicePid =
+                ((ServiceEvent) event.getProperty(EventConstants.EVENT)).getServiceReference()
+                        .getProperty(Constants.SERVICE_PID);
+        if (servicePid == null) {
+            return;
+        }
+
+        RegistryStore service = registryStorePidToServiceMap.get(servicePid);
+        if (service == null) {
+            return;
+        }
+        executor.execute(() -> {
+            String registryId = service.getRegistryId();
+            try {
+                List<Metacard> metacards =
+                        Security.runAsAdminWithException(() -> federationAdminService.getInternalRegistryMetacardsByRegistryId(
+                                registryId));
+                List<String> idsToDelete = metacards.stream()
+                        .map(Metacard::getId)
+                        .collect(Collectors.toList());
+                if (!CollectionUtils.isEmpty(idsToDelete)) {
```

Comment:
Collectors never return `null`, so `!idsToDelete.isEmpty()` should be good enough here.